avogadro/core/residue.cpp

```diff
@@ -0,0 +1,129 @@
+/******************************************************************************
+
+  This source file is part of the Avogadro project.
+
+  Copyright 2018 Kitware, Inc.
+
+  This source code is released under the New BSD License, (the "License").
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+******************************************************************************/
+
+#include "residue.h"
+#include "molecule.h"
+#include "residuedata.h"
+
+#include <algorithm>
+#include <cassert>
+
+namespace Avogadro {
+namespace Core {
+
+Residue::Residue() {}
+
+Residue::Residue(std::string& name)
+  : m_residueName(name)
+{}
+
+Residue::Residue(std::string& name, Index& number)
+  : m_residueName(name)
+  , m_residueId(number)
+{}
+
+Residue::Residue(std::string& name, Index& number, char& id)
+  : m_residueName(name)
+  , m_residueId(number)
+  , m_chainId(id)
+{}
+
+Residue::Residue(const Residue& other)
+  : m_residueName(other.m_residueName)
+  , m_residueId(other.m_residueId)
+  , m_atomNameMap(other.m_atomNameMap)
+{}
+
+Residue& Residue::operator=(Residue other)
+{
+  using std::swap;
+  swap(*this, other);
+  return *this;
+}
+
+Residue::~Residue() {}
+
+std::string Residue::residueName()
+{
+  return m_residueName;
+}
+
+void Residue::setResidueName(std::string& name)
+{
+  m_residueName = name;
+}
+
+Index Residue::residueId()
+{
+  return m_residueId;
+}
+
+void Residue::setResidueId(Index& number)
+{
+  m_residueId = number;
+}
+
+char Residue::chainId()
+{
+  return m_chainId;
+}
+
+void Residue::setChainId(char& id)
+{
+  m_chainId = id;
+}
+
+void Residue::addResidueAtom(std::string& name, Atom& atom)
+{
+  m_atomNameMap.insert(std::pair<std::string, Atom>(name, atom));
+}
+
+std::vector<Atom> Residue::residueAtoms()
+{
+  std::vector<Atom> res;
+  for (AtomNameMap::iterator it = m_atomNameMap.begin();
+       it != m_atomNameMap.end(); ++it) {
+    res.push_back(it->second);
+  }
+  return res;
+}
+
+void Residue::resolveResidueBonds(Molecule& mol)
```

Comment:
I'm not sure this best lives here. I mean, the code is okay, but if it's only in the Residue class, then to process a PDB file, you need to call this each time you finish a residue? Wouldn't you rather have a Molecule call or a separate ResidueBondPerception class that handles the whole molecule at once?

Reply:
Agreed. So, instead of calling the method upon every new residue, how about modifying the Molecule class to store an array of residues and invoke a function such that once the reading is done, call the above method for each and every residue instance from the residue array?

Reply:
Yes, exactly. Clearly the Molecule class needs to track the array of residues (if present).